HllSketch Merge/Build BufferAggregators: Speed up init with prebuilt sketch.

gianm opened the pull request:
Inspired by the following flame graph, this patch attacks `HllSketchMergeBufferAggregator.init` by precomputing a single initialized sketch, and then copying that precomputed sketch into aggregation buffers. Before, this was done by calling `Unsafe.setMemory` (by way of `Memory.clear`).

<img width="1655" alt="image" src="https://user-images.githubusercontent.com/1214075/62092710-0e042000-b22b-11e9-9385-54b6cd7d1de2.png">

One interesting thing about this flame graph is that `Memory.clear` accounts for nearly 60% (!) of it. Of that, half is in the HllSketch constructor, which this patch attacks.

The rest is mostly `Union.update`, in particular two methods it calls: `DirectCouponHashSet.growHashSet`, and `DirectCouponList.promoteListToSet`. @leerho or @AlexanderSaydakov - for these two methods - any idea what size of memory would typically be cleared at once? We could potentially speed these up by using copyMemory instead of setMemory with some pre-zeroed chunks.

JMH benchmarks (~5x speedup in `DataSketchesBenchmark.init`):

```
master

Benchmark                            Mode  Cnt         Score        Error  Units
DataSketchesBenchmark.init          thrpt   15   4788022.652 ± 109374.009  ops/s
DataSketchesBenchmark.initAndGet    thrpt   15   4085628.980 ±  86160.850  ops/s
DataSketchesBenchmark.initAndSerde  thrpt   15   2598419.066 ± 126836.466  ops/s

master + init change

Benchmark                            Mode  Cnt         Score        Error  Units
DataSketchesBenchmark.init          thrpt   15  24943146.877 ± 140698.560  ops/s
DataSketchesBenchmark.initAndGet    thrpt   15  12141905.292 ±  99566.648  ops/s
DataSketchesBenchmark.initAndSerde  thrpt   15   4920001.015 ±  62927.010  ops/s
```

AlexanderSaydakov commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchMergeBufferAggregator.java:
```diff
@@ -60,17 +63,22 @@ public HllSketchMergeBufferAggregator(
+    this.emptySketch = new byte[size];
+
+    //noinspection ResultOfObjectAllocationIgnored (Union writes to "emptySketch" as a side effect of construction)
+    new Union(lgK, WritableMemory.wrap(emptySketch, ByteOrder.LITTLE_ENDIAN).writableRegion(0, size));
```
no need to call writableRegion here since the region is identical to the whole thing

gianm replied:
Thanks, I removed this.

leerho commented:
@gianm 
Thank you for this analysis!  The current HllSketch starts at a very small size and grows in size by factors of 2 until it is just under the size of the dense sketch and then converts to dense mode, after which it stays fixed.  See the bottom chart [here](https://datasketches.github.io/docs/HLL/HLL.html).

I have thought about an option that would eliminate the sparse mode.  An empty sketch would be created at full size.  Here is where we might be able to take advantage of a high-speed copy of pre-initialized memory upon sketch creation.  For the life of the sketch it would never need to allocate or clear any memory.  (If you are using HLL-4 mode, there is a little lookup cache that occasionally may need to resize, but that should be rare.)

Druid (so far) always allocates space for the maximum size sketch and cannot take advantage of the fact that the size of the sketch for small cardinalities can be quite small and then grow.  If this is going to remain true, then it might make sense to add this option.   Druid is wasting a lot of memory because of the fixed allocation sizes and this has been discussed at length in other threads, which I'm not going to repeat here.  Nonetheless, I am likely out-of-date with what the latest thinking within the Druid team is on this topic, so please update me :) 

Clearly our focus has been to optimize the best possible accuracy for a given size.  But if you are willing to allow the sketch to start at maximum size this may make sense for you.  It could mean a substantial speed up. When implemented, it will be a configuration option, so you could always play with it and change your mind.

The downside of eliminating the sparse mode is that the accuracy of the sketch for low cardinalities will be worse.  It will be within the error guarantees for a given _k_ and behave like the conventional Flajolet-Martin HLL sketch at low cardinalities.  Our HLL sparse mode takes advantage of new estimators that were developed for our CPC sketch which outperforms the accuracy / byte of almost any other HLL sketch out there by a large margin.

We are in the middle of this Apache migration, so implementing this may be some months out, unless we could get some help speeding up the migration :) :)

Let me know if this makes sense to you.

AlexanderSaydakov commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchMergeBufferAggregator.java:
```diff
@@ -40,14 +40,17 @@
   private final TgtHllType tgtHllType;
   private final int size;
   private final Striped<ReadWriteLock> stripedLock = Striped.readWriteLock(NUM_STRIPES);
+  private final byte[] emptySketch;
```
this variable should better be called emptyUnion

gianm replied:
Renamed it.

AlexanderSaydakov commented:
I would think that a similar change should be made to HllSketchBuildBufferAggregator

gianm commented:
> I would think that a similar change should be made to HllSketchBuildBufferAggregator

I just added that too, and updated the PR title accordingly.

Btw, I couldn't share code for creating the prebuilt sketches, because init does not behave the same way in the two flavors. The "build" flavor initializes based on tgtHllType, but the "merge" flavor always initializes using HLL_8. Do you know why that is?

gianm commented:
> Druid (so far) always allocates space for the maximum size sketch and cannot take advantage of the fact that the size of the sketch for small cardinalities can be quite small and then grow. If this is going to remain true, then it might make sense to add this option. Druid is wasting a lot of memory because of the fixed allocation sizes and this has been discussed at length in other threads, which I'm not going to repeat here. Nonetheless, I am likely out-of-date with what the latest thinking within the Druid team is on this topic, so please update me :)

The most recent discussion has been happening in here: #8126. I am not sure what form yet resizable aggregators will take but let's assume they will happen!

> Clearly our focus has been to optimize the best possible accuracy for a given size. But if you are willing to allow the sketch to start at maximum size this may make sense for you. It could mean a substantial speed up. When implemented, it will be a configuration option, so you could always play with it and change your mind.

It definitely sounds like a useful option, especially if the accuracy decline isn't too bad. Maybe some hybrid approach would also be useful — starting at some medium point rather than very small or fully dense?

> The downside of eliminating the sparse mode is that the accuracy of the sketch for low cardinalities will be worse. It will be within the error guarantees for a given _k_ and behave like the conventional Flajolet-Martin HLL sketch at low cardinalities. Our HLL sparse mode takes advantage of new estimators that were developed for our CPC sketch which outperforms the accuracy / byte of almost any other HLL sketch out there by a large margin.

Any idea how much worse?

> We are in the middle of this Apache migration, so implementing this may be some months out, unless we could get some help speeding up the migration :) :)

I joined the DataSketches dev list a few days ago, anything I could do to help?

AlexanderSaydakov commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchBuildBufferAggregator.java:
```diff
@@ -53,6 +55,14 @@
+   * {@link HllSketchMergeBufferAggregator} does something similar, but we don't share code to create emptySketch
+   * because it is not exactly the same (the "build" flavor initializes based on tgtHllType, but the "merge" flavor
+   * always initializes using HLL_8).
+   */
```
This is not so. HllSketchMergeBufferAggregator initializes HllUnion objects as opposed to HllSketch objects here. And HllUnion constructor does not even have the target type parameter (it is always in HLL_8 mode, but it is an implementation detail). The target type is used when the final result is obtained from the union in get().

gianm replied:
Thanks for clarifying. I just read the javadocs for Union and HllSketch and the situation is more clear. I adjusted the comment to the following,

```java
  /**
   * Used by {@link #init(ByteBuffer, int)}. We initialize by copying a prebuilt empty Union image.
   * {@link HllSketchBuildBufferAggregator} does something similar, but different enough that we don't share code. The
   * "build" flavor uses {@link HllSketch} objects and the "merge" flavor uses {@link Union} objects.
   */
```

AlexanderSaydakov commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchMergeBufferAggregator.java:
```diff
@@ -60,17 +70,25 @@ public HllSketchMergeBufferAggregator(
+    this.emptyUnion = new byte[size];
+
+    //noinspection ResultOfObjectAllocationIgnored (Union writes to "emptySketch" as a side effect of construction)
+    new Union(lgK, WritableMemory.wrap(emptyUnion, ByteOrder.LITTLE_ENDIAN));
```
No need to force byte order when wrapping byte[]

gianm replied:
Removed it.

AlexanderSaydakov commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchMergeBufferAggregator.java:
```diff
@@ -60,17 +69,25 @@ public HllSketchMergeBufferAggregator(
-    // The objects are not cached as in BuildBufferAggregator since they never exceed the max size and never move.
-    // So it is easier to reconstruct them by wrapping memory then to keep position-to-object mappings. 
-    new Union(lgK, mem);
+    // Copy prebuilt empty sketch object.
```
should be union here

gianm replied:
Ah good catch. Updated this one too.

AlexanderSaydakov commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchMergeBufferAggregator.java:
```diff
@@ -60,17 +69,25 @@ public HllSketchMergeBufferAggregator(
   {
-    final WritableMemory mem = WritableMemory.wrap(buf, ByteOrder.LITTLE_ENDIAN).writableRegion(position, size);
-    // Not necessary to keep the constructed object since it is cheap to reconstruct by wrapping the memory.
-    // The objects are not cached as in BuildBufferAggregator since they never exceed the max size and never move.
```
Perhaps it is worth keeping this comment explaining why there is no caching of union objects is done as opposed to cashing of sketches in the HllSketchBuildBufferAggregator

gianm replied:
Agreed, added it back. I put it in the "aggregate" method since that seems to be the most useful place to put it (it's where Union objects are most commonly created).

AlexanderSaydakov replied:
In aggregate() they are updated, but constructed in init()
The comment is about why "new" was called and the result not kept, and also why there is no caching like in the build aggregator.

gianm replied:
But after this patch, they aren't constructed in init any longer. So I thought including this comment in "init" would be confusing to a reader, and "aggregate" was the next-best place (since that's now where Union objects are created and thrown away after being used):

```java
    try {
      // Not necessary to keep the constructed object since it is cheap to reconstruct by wrapping the memory.
      // The objects are not cached as in HllSketchBuildBufferAggregator, since they never exceed the max size and
      // never move.
      final Union union = Union.writableWrap(mem);
      union.update(sketch);
    }
```

I suppose the comment could go into the constructor instead, which is where the "new Union" code from "init" migrated. But the comment, as currently written, is a non-sequitur there, since there would be no reason to cache the Union object in the constructor anyway (it's only used to generate the emptyUnion bytes). It would need to be rewritten somehow.

I think things make sense the way they are in the latest patch, but if you have a better suggestion I am all ears.

AlexanderSaydakov commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchBuildBufferAggregator.java:
```diff
@@ -64,11 +74,27 @@ public HllSketchBuildBufferAggregator(
+      buf.position(oldPosition);
+    }
+
+    // Add an HllSketch for this chunk to our sketchCache.
```
ops, the above buf.put() has no effect because here a new one is constructed in that memory region and also cached

gianm replied:
Ah, thanks for the catch. I replaced it with:

```java
    putSketchIntoCache(buf, position, HllSketch.writableWrap(mem));
```

AlexanderSaydakov commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchBuildBufferAggregator.java:
```diff
@@ -64,11 +74,27 @@ public HllSketchBuildBufferAggregator(
+
+    // Add an HllSketch for this chunk to our sketchCache.
     final WritableMemory mem = getMemory(buf).writableRegion(position, size);
     putSketchIntoCache(buf, position, new HllSketch(lgK, tgtHllType, mem));
```
I think this should be:
putSketchIntoCache(buf, position, HllSketch.writableWrap(mem));

gianm replied:
That does look right, thanks for the catch. Updated.

AlexanderSaydakov commented:
Looks good to me.
I would leave that comment in the init() though. Perhaps rephrase it a little. The objects are still "constructed" there by copying an empty union.

gianm commented:
I rewrote the comment and added it to "init":

```
    // Copy prebuilt empty union object.
    // Not necessary to cache a Union wrapper around the initialized memory, because:
    //  - It is cheap to reconstruct by re-wrapping the memory in "aggregate" and "get".
    //  - Unlike the HllSketch objects used by HllSketchBuildBufferAggregator, our Union objects never exceed the
    //    max size and therefore do not need to be potentially moved in-heap.
```

I believe it is accurate.

BTW, I removed the "never move" language because I believe it isn't accurate. "relocate" can still be called on the merge flavor -- it happens when e.g. the groupBy engine is resizing its internal hash table and moving objects around as a result. But the merge flavor shouldn't need to do anything special on a relocation, since it doesn't have any position based caches.

leerho commented:
@gianm 
You offered:

> I joined the DataSketches dev list a few days ago, anything I could do to help?

Rather than muddying up this thread I have posted a [Migration Help] request on dev@datasketches.apache.org.  Assuming that is the "proper" place to put such a request.  I considered creating an issue on Memory, but thought I would try the "Apache way" first :).

We can try dev@, but if it gets too messy we can move to a GitHub issue.

Thanks!

AlexanderSaydakov commented:
looks good to me

gianm commented:
Thank you for the review!
